xtask/src/commands/test.rs

```diff
@@ -28,6 +29,23 @@ impl Test {
             "--no-demo and --with-demo are mutually exclusive",
         );
 
+        // start building tests, while the subservices are spinning up
+        eprintln!("Starting to build tests...");
+        let _build = std::process::Command::new(which::which("cargo")?)
```

Comment:
1. You don't need to store the result in a variable. Dropping it won't kill the process.
2. On the other hand: you should wait on your process right after making sure the node services are ready. You should also add a log that says "Waiting for background process that pre-compiles the test to finish..."
3. You should also redirect stdout and stderr to null to avoid getting the logs mixed up. Do no report the exit status of that process either. It's okay to lose the logs here because the tests commands will be run again after anyway and show any eventual error.
4. This background thingy should be behind a flag. You can add a flag to the cli (line 15). The feature must be disabled by default (undesirable in most cases) and enabled in the CI's yaml file.

Reply:
I don't understand why the background building should be disabled by default, would we rather be idle waiting for the npm steps to complete?

Reply:
edit:

2 - running cargo test in the same directory will trigger `Blocking waiting for file lock on build directory` until the `--no-run` command is complete, which behaves the same as .join() in the end. I'm fine either way.  The nice thing about joining is that we can print messages tho.

Reply:
> However I don't understand why the background building should be disabled by default, would we rather be idle waiting for the npm steps to complete?

Locally the things are ready most of the time. It's not ready only the first time. Plus users who will use `--no-demo` really don't need this so you can even make the flags conflicting.

> 2 - running cargo test in the same directory will trigger Blocking waiting for file lock on build directory until the --no-run command is complete, which behaves the same as .join() in the end. I'm fine either way. The nice thing about joining is that we can print messages tho.

That was to be expected. That is why I prefer that we wait explicitly (also to avoid zombie processes). You can print "Waiting for bg task to complete..." and you can print "Bg test failed/succeeded" but it must not be fatal. Since the logs will be disabled, the user needs to see the compiler errors. For that the tests should run even if the bg task failed and that will show the errors to the user.

Reply:
lgtm, let's only spawn the background build task when running `--with-demo` then

Reply:
Wait no. Don't re-use the flag!

You want to explain in the cli's doc "Run the federation demo in bg... oh btw, it runs something else in background" ? :joy:

Reply:
> Locally the things are ready most of the time. It's not ready only the first time.

running with demo adds an npm i step, which takes roughly 30 seconds on my machine, regardless of whether node_modules are up to date or not (which is afaiu what you meant by `locally the things are ready most of the time` ?). This is the reason why I spawned this task in the background.

As for the documentation I would probably go with `run the federation demo in background, xtask will prepare the next steps while federation demo is being started` or something ?

Reply:
Your machine is slow.
```
Executed in    2.70 secs    fish           external
   usr time    3.91 secs  370.00 micros    3.91 secs
   sys time    0.28 secs   55.00 micros    0.28 secs
```

Anyway. My point is that a cli parameter should be explicit. `--with-demo` should only mean "run the federation demo in background" and not "run the federation demo in background and start another background process to optimize your time". I prefer to have that under a separate parameter like `--pre-compile`.

Reply:
fine by me, I don't use xtask a lot anyway, lets go with that

Reply:
well you would use `--no-demo` anyway right?

Reply:
uh i would cargo xtask test and 🤞 😅

Reply:
it's the default anyway so yeah just xtask test and :crossed_fingers: